src/Interpreters/AsynchronousMetrics.cpp

```diff
@@ -329,6 +325,48 @@ void AsynchronousMetrics::update()
     saveAllArenasMetric<size_t>(new_values, "muzzy_purged");
 #endif
 
+    // Try to add processor frequencies, ignoring errors.
+    try
+    {
+        ReadBufferFromFile buf("/proc/cpuinfo", 32768 /* buf_size */);
+
+        // We need the following lines:
+        // core id : 4
+        // cpu MHz : 4052.941
+        // They contain tabs and are interspersed with other info.
+        int core_id = 0;
+        while (!buf.eof())
+        {
+            std::string s;
+            // We don't have any backslash escape sequences in /proc/cpuinfo, so
+            // this function will read the line until EOL, which is exactly what
+            // we need.
+            readEscapedStringUntilEOL(s, buf);
+            // It doesn't read the EOL itself.
+            ++buf.position();
+
+            if (s.rfind("core id", 0) == 0)
```

Comment:
This assumes single CPU per system, because core id is unique per physical CPU. I'm also not sure if hyperthreads necessarily run at the same frequency on all CPUs. Having a metric per logical CPU alleviates both issues.